### PR TITLE
[training, perf] feat: Add eval-time context parallelism via decentralized PG rebinding

### PR DESCRIPTION
**Pull request by cuichenx**
## Summary

Enables validation to run with a **higher CP degree than training**, addressing the wasted-work problem when DP_train is large and the validation set is small. The mechanism swaps cached process-group references on every module from `train_pgs` to `eval_pgs` for the duration of `evaluate()`, then restores them on exit. Works only with the **decentralized process-group path** (`use_decentralized_pg=True`); no Megatron-Core changes required.

### Motivation

For an 8192-GPU run with TP=2, PP=4, CP_train=1 (so DP_train=1024) and a 1024-sample validation set, you must set GA=4 to keep the PP pipeline full, processing 4096 sample-steps — **4× redundant work** vs evaluating each sample once. With CP_eval=4, DP_eval=256 covers all 1024 samples in one iter with no redundancy and ~CP_eval× lower activation memory.

## What's added

- **`src/megatron/bridge/training/eval_cp.py`** — `install_pg_collection()` walks the module tree and rebinds:
  - `pg_collection` (used by `TransformerLayer`, `DotProductAttention`, …)
  - Named CP-bearing group attributes (`cp_group`, `tp_cp_group`, `tp_dp_cp_group`, `dp_cp_group`)
  - `TEDotProductAttention` internal CP comm state via `set_context_parallel_group()`

  `RotaryEmbedding` / `MultimodalRotaryEmbedding` instances are intentionally skipped — see the in-code comment for the double-shard bug this avoids. The `@lru_cache` on `forward` is cleared defensively.

  `eval_cp_context()` is a context manager wrapping install/restore.

- **`src/megatron/bridge/training/config.py`** — `DistributedInitConfig.eval_context_parallel_size: int | None = None`
- **`src/megatron/bridge/training/state.py`** — `GlobalState.train_pgs` / `.eval_pgs` (`ProcessGroupCollection | None`)
- **`examples/decentralized_pg/pretrain_qwen3_eval_cp.py`** — 4-GPU end-to-end demo on Qwen3-4B (TP=2, CP_train=1, CP_eval=2): training → baseline eval at CP=1 → eval-CP at CP=2 → verify-restore at CP=1.

## Constraints

- Requires `dist.use_decentralized_pg=True`
- No FSDP / CUDA graphs / hierarchical CP today
- `seq_length % (2 * cp_eval) == 0`

## Test plan

- [x] 4-GPU manual run (`TP=2 CP_train=1 CP_eval=2`, Qwen3-4B): train 3 iters, baseline eval CP=1 succeeds (loss 7.870), eval-CP CP=2 succeeds (loss 8.836), verify-restore eval CP=1 succeeds (loss 7.978)
- [x] `uv run pre-commit run` passes (ruff, ruff-format)
- [ ] Add a unit test for `install_pg_collection` rebinding semantics
- [ ] Add a 2-GPU functional test (within CI budget) covering the eval-CP path
- [ ] Multi-node benchmark vs the original 8K-GPU motivating scenario

## Notes for reviewers

- `RotaryEmbedding.cp_group` deliberately *not* updated. With `pre_process` SP scatter at eval, the embedding is sharded across `tp*cp_eval` so `decoder_input.size(0) = seq/(tp*cp_eval)`. After QKV's TP AllGather, query is `[seq/cp_eval, …]`. `get_rotary_seq_len` multiplies by tp_size and the *config*'s cp_size (= cp_train), so freqs land at `seq/cp_eval` *before* CP slicing. If we then slice freqs by `cp_eval` we'd double-shard them — concrete repro shows query `(512,…)` vs freqs `(256,…)` mismatch in `_apply_rotary_pos_emb_bshd`. Skipping the rotary cp_group update keeps freqs at `seq/cp_eval`, matching query.
- The example uses Qwen3 with YaRN RoPE end-to-end, which exercises the unfused `_apply_rotary_pos_emb_bshd` path; the same logic also applies to standard RoPE.

### CODE REVIEW

File: src/megatron/bridge/training/state.py [hunk 1]
```diff
@@ -149,6 +151,10 @@ def __init__(self) -> None:
         self._nvrx_straggler_created: bool = False
         self._energy_monitor: Optional[Any] = None
         self._energy_monitor_created: bool = False
+        # Eval-time context parallelism: set by the caller when eval_context_parallel_size
+        # differs from the training CP degree. eval_cp.eval_cp_context reads these.
+        self.train_pgs: Optional[ProcessGroupCollection] = None
```
Comment (yaoyu-33) [inline src/megatron/bridge/training/state.py]: make them private?

File: src/megatron/bridge/training/eval_context_parallel_rebinding.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
+# Copyright (c) 2026, NVIDIA CORPORATION.  All rights reserved.
```
Comment (yaoyu-33) [inline src/megatron/bridge/training/eval_context_parallel_rebinding.py:1]: naming eval_cp sounds a bit vague, can we come up with a new name.

This seems not in PoR.

Comment (cuichenx) [reply src/megatron/bridge/training/eval_context_parallel_rebinding.py:1]: i changed it to eval_context_parallel_rebinding. It's in the roadmap issue already~

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3755/commits/e5c61851ed23db1cc41b6db62facfb90670c296a

**Review by yaoyu-33 — APPROVED** [review]
